src/job_cache/daemon_cache.cpp

```diff
@@ -0,0 +1,1324 @@
+
+  JAST input_dirs_json(JSON_ARRAY);
+  for (const auto &input_dir : directories) {
+    input_dirs_json.add("", input_dir.to_json());
+  }
+  json.add("input_dirs", std::move(input_dirs_json));
+
+  JAST output_files_json(JSON_ARRAY);
+  for (const auto &output_file : outputs) {
+    output_files_json.add("", output_file.to_json());
+  }
+  json.add("output_files", std::move(output_files_json));
+
+  JAST output_directories_json(JSON_ARRAY);
+  for (const auto &output_directory : output_dirs) {
+    output_directories_json.add("", output_directory.to_json());
+  }
+  json.add("output_dirs", std::move(output_directories_json));
+
+  JAST output_symlinks_json(JSON_ARRAY);
+  for (const auto &output_symlink : output_symlinks) {
+    output_symlinks_json.add("", output_symlink.to_json());
+  }
+  json.add("output_symlinks", std::move(output_symlinks_json));
+
+  return json;
+}
+
+FindJobRequest::FindJobRequest(const JAST &find_job_json) {
+  cwd = find_job_json.get("cwd").value;
+  command_line = find_job_json.get("command_line").value;
+  envrionment = find_job_json.get("envrionment").value;
+  stdin_str = find_job_json.get("stdin").value;
+
+  // Read the input files, and compute the directory hashes as we go.
+  for (const auto &input_file : find_job_json.get("input_files").children) {
+    std::string path = input_file.second.get("path").value;
+    Hash256 hash = Hash256::from_hex(input_file.second.get("hash").value);
+    bloom.add_hash(hash);
+    visible[std::move(path)] = hash;
+  }
+
+  // Now accumulate the hashables in the directory.
+  std::unordered_map<std::string, std::string> dirs;
+  // NOTE: `visible` is already sorted because its an std::map.
+  // this means that we'll accumulate directories correctly.
+  for (const auto &input : visible) {
+    auto pair = parent_and_base(input.first);
+    if (!pair) continue;
+    std::string parent = std::move(pair->first);
+    std::string base = std::move(pair->second);
+    dirs[parent] += base;
+    dirs[parent] += ":";
+  }
+
+  // Now actually perform those hashes
+  for (auto dir : dirs) {
+    dir_hashes[dir.first] = Hash256::blake2b(dir.second);
+  }
+
+  // When outputting files we need to map sandbox dirs to output dirs.
+  // Collect those redirects here.
+  for (const auto &dir_redirect : find_job_json.get("dir_redirects").children) {
+    auto dir_range = wcl::make_filepath_range(dir_redirect.first);
+    dir_redirects.move_emplace(dir_range.begin(), dir_range.end(), dir_redirect.second.value);
+  }
+}
+
+JAST FindJobRequest::to_json() const {
+  JAST json(JSON_OBJECT);
+  json.add("cwd", cwd);
+  json.add("command_line", command_line);
+  json.add("envrionment", envrionment);
+  json.add("stdin", stdin_str);
+
+  JAST input_files(JSON_ARRAY);
+  for (const auto &input_file : visible) {
+    JAST input_entry(JSON_OBJECT);
+    input_entry.add("path", input_file.first);
+    input_entry.add("hash", input_file.second.to_hex());
```

Comment:
I think this can just be `input_file.to_json()`

Reply:
I think the issue is that `visible` is a map and so when you iterate over it you get pairs. `visible` is a map for efficiency and sorting reasons. We could probably make it an std::vector, sort the std vector, and then do lookups via binary search if you felt strongly about this though.

Reply:
Nah, it was an offhanded comment assuming there was an underlying struct. That assumption was wrong so no reason for this comment